### PR TITLE
Release resources acquired in `prepareRequest`

### PR DESCRIPTION
**Pull request by DaveCTurner**
Some `BaseRestHandler#prepareRequest` implementations may construct a
request with nontrivial lifecycle (e.g. containing a `SecureString`),
but there's no hook to clean up on an exception: the handler only
continues with execution if the REST request was valid. This commit adds
a `Releasable` superinterface to `RestChannelConsumer` whose `close()`
method is invoked whether the action executes or not.

### CODE REVIEW

**Comment by DaveCTurner** [comment]
@masseyke I suspect you might need something like this for https://github.com/elastic/elasticsearch/pull/103356 too.

**Review by mosche — COMMENTED** [review]
Just raising a slight concern here, the implemented behavior might be confusing / prone to lifecycle errors which might not easily surface in single threaded unit tests.
Channels are commonly consumed (written to) async through listeners. Though, by then the `RestChannelConsumer` itself might already be closed. I'm not sure if this is an issue for the use cases you have in mind, but it might warrent some careful documentation.

**Comment by DaveCTurner** [comment]
> Channels are commonly consumed (written to) async through listeners. Though, by then the RestChannelConsumer itself might already be closed.

That's ok tho, the channel itself will live on. I added a note in 551566a48f5, does that help?

File: server/src/main/java/org/elasticsearch/rest/BaseRestHandler.java [hunk 1]
```diff
@@ -77,30 +78,33 @@ public final long getUsageCount() {
...
+            if (request.hasContent() && request.isContentConsumed() == false) {
+                throw new IllegalArgumentException(
+                    "request [" + request.method() + " " + request.path() + "] does not support having a body"
+                );
+            }
 
-        usageCount.increment();
-        // execute the action
-        action.accept(channel);
+            usageCount.increment();
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/rest/BaseRestHandler.java:105]: It's a success count. We could count invalid requests as well (maybe separately) but that's a task for another day.

**Review by mosche — APPROVED** [review]
> > Channels are commonly consumed (written to) async through listeners. Though, by then the RestChannelConsumer itself might already be closed.
> 
> That's ok tho, the channel itself will live on. I added a note in [551566a](https://github.com/elastic/elasticsearch/commit/551566a48f5190d5296f59c62c6904ea1628f8f0), does that help?

Perfect, thanks for adding the docs 👍 
LGTM

**Pull request merged**